worker/uniter/runner/context/context.go

```diff
@@ -1302,8 +1309,45 @@ func (ctx *HookContext) doFlush(process string) error {
 		b.AddStorage(ctx.storageAddConstraints)
 	}
 
-	b.AddSecretCreates(ctx.secretChanges.pendingCreates)
-	b.AddSecretUpdates(ctx.secretChanges.pendingUpdates)
+	// Before saving the secret metadata to Juju, save the content to an external
+	// store (if configured) - we need the provider id to send to Juju.
+	// If the flush to Juju fails, we'll delete the external content.
+	var cleanups []string
+	pendingCreates := make([]uniter.SecretCreateArg, len(ctx.secretChanges.pendingCreates))
+	pendingUpdates := make([]uniter.SecretUpsertArg, len(ctx.secretChanges.pendingUpdates))
+	for i, c := range ctx.secretChanges.pendingCreates {
+		providerId, err := ctx.secrets.SaveContent(c.URI, 1, c.Value)
+		if errors.IsNotSupported(err) {
+			pendingCreates[i] = c
+			continue
+		}
+		if err != nil {
+			return errors.Annotatef(err, "saving content for secret %q", c.URI.ID)
+		}
+		cleanups = append(cleanups, providerId)
+		c.ProviderId = &providerId
+		c.Value = nil
+		pendingCreates[i] = c
+	}
+	for i, u := range ctx.secretChanges.pendingUpdates {
+		// Juju checks that the current revision is stable when updating metadata so it's
+		// safe to increment here knowing the same value will be saved in Juju.
+		providerId, err := ctx.secrets.SaveContent(u.URI, u.CurrentRevision+1, u.Value)
```

Comment:
Does SaveContent have a guarantee it won't overwrite a value (i.e. create only)?

Reply:
We ensure the revision is incremented - it's up to the secret store implementation to create content specific to that revision. The k8s implementation uses "id-revison" as the secret name so will return already exists error if there's a coding mistake